winch/codegen/src/visitor.rs

```diff
@@ -705,6 +680,217 @@ where
+                        vmctx.into(),
+                        dst.try_into().unwrap(),
+                        src.try_into().unwrap(),
+                    ]
+                    .into_iter(),
+                );
+                call.calculate_call_stack_space(cx).reg(masm, cx, callee);
+            },
+        )
+    }
+
+    fn visit_table_get(&mut self, table: u32) {
+        let ptr_type = self.env.ptr_type();
+        let table_index = TableIndex::from_u32(table);
+        let table_data = self.env.resolve_table_data(table_index);
+        let plan = self.env.table_plan(table_index);
+        let heap_type = plan.table.wasm_ty.heap_type;
+        let style = plan.style.clone();
+        let table_get = self
+            .env
+            .builtins
+            .table_get_lazy_init_func_ref::<M::ABI, M::Ptr>();
+
+        FnCall::new(&table_get.sig).with_lib(
+            self.masm,
+            &mut self.context,
+            &table_get,
+            |cx, masm, call, callee| {
+                match heap_type {
+                    WasmHeapType::Func => match style {
+                        TableStyle::CallerChecksSignature => {
+                            CodeGen::emit_lazy_init_funcref(
+                                &table_data,
+                                table_index,
+                                ptr_type,
+                                cx,
+                                masm,
+                                call,
+                                callee,
+                            );
+                        }
+                    },
+                    t => unimplemented!("Support for WasmHeapType: {t}"),
+                };
+            },
+        );
+    }
+
+    fn visit_table_grow(&mut self, table: u32) {
+        let ptr_type = self.env.ptr_type();
+        let table_index = TableIndex::from_u32(table);
+        let table_plan = self.env.table_plan(table_index);
+        let vmctx = TypedReg::new(ptr_type, <M::ABI as ABI>::vmctx_reg());
+        let builtin = match table_plan.table.wasm_ty.heap_type {
+            WasmHeapType::Func => self.env.builtins.table_grow_func_ref::<M::ABI, M::Ptr>(),
+            ty => unimplemented!("Support for HeapType: {ty}"),
+        };
+
+        FnCall::new(&builtin.sig).with_lib(
+            self.masm,
+            &mut self.context,
+            &builtin,
+            |cx, masm, call, callee| {
+                let len = cx.stack.len();
+                // table.grow requires at least 2 elements on the value stack.
+                debug_assert!(len >= 2);
+                // The table_grow builtin expects the parameters in a different
+                // order.
+                // The value stack at this point should contain:
+                // [ init_value | delta ] (stack top)
+                // but the builtin function expects the init value as the last
+                // argument.
+                cx.stack.inner_mut().swap(len - 1, len - 2);
```

Comment:
I'm wondering if there's a chance we could make the signatures of the libcalls compatible with the Wasm spec, that way we would avoid having to pre-process the stack before emitting calls.

Reply:
That seems like a reasonable change to me, @fitzgen is there any reason not to match the spec signature?

Reply:
https://github.com/bytecodealliance/wasmtime/issues/7179